EduVPN/ConfigurationService.swift

```diff
@@ -40,13 +75,64 @@ class ConfigurationService {
         }
     }
     
-    private func restoreOrCreateKeyPair(for profile: Profile, authState: OIDAuthState, handler: @escaping (Either<(certificate: String, privateKey: String)>) -> ()) {
-        // TODO: restore key pair
-        createKeyPair(for: profile, authState: authState, handler: handler)
+    /// Checks if keypair is available for provider, otherwise creates and stores new keypair
+    ///
+    /// - Parameters:
+    ///   - info: Provider info
+    ///   - authState: Authencation token
+    ///   - handler: Keypair or error
+    private func restoreOrCreateKeyPair(for info: ProviderInfo, authState: OIDAuthState, handler: @escaping (Either<(certificate: String, privateKey: String)>) -> ()) {
+        // TODO: restore key pair from keychain instead of user defaults
+        var keyPairs = UserDefaults.standard.array(forKey: "keyPairs") ?? []
+        let keyPair = keyPairs.first { (keyPair) -> Bool in
+            guard let keyPair = keyPair as? [String: AnyObject] else {
+                return false
+            }
+            
+            guard let providerBaseURL = keyPair["providerBaseURL"] as? String, info.provider.baseURL.absoluteString == providerBaseURL else {
+                return false
+            }
+            
+            guard let _ = keyPair["certificate"] as? String else {
+                return false
+            }
+            
+            guard let _ = keyPair["privateKey"] as? String else {
+                return false
+            }
+            
+            return true
+        }
+        
+        if let keyPair = keyPair as? [String: AnyObject] {
+            let certificate = keyPair["certificate"] as! String
+            let privateKey = keyPair["privateKey"] as! String
+            handler(.success((certificate: certificate, privateKey: privateKey)))
```

Comment:
Both of these `as!` instances are clearly safe looking at the code right above, but in case you want to avoid having to use them regardless:

Rather than doing `keyPairs.first(where:)`, you can use `keyPairs.lazy.flatMap` instead, with a closure body of `keyPair -> (certificate: String, privateKey: String)?`. In the body you'd replace the two instances of `guard let _ =` by `guard let certificate =` and `guard let privateKey =`, the instances of `return false` become `return nil` and `return true` becomes `return (certificate, privateKey)`. The result is an optional certificate-privateKey pair of non-optional strings, so you don't end up doing each lookup twice.

And the `lazy` is there to prevent unnecessarily many iterations. 🙂